### DISCUSSION TITLE
Generic Components

### DISCUSSION THREAD

**Discussion by gaku-sei** [mint-lang/mint#130] [Language Development]
According to this line: https://github.com/mint-lang/mint/blob/master/src/messages/property_not_hole.cr#L9, "type hole" or "generics" are not yet supported in properties (neither are they supported in records).

But they are "partly" supported in functions, since one can write:

```
fun getFirstItem (xs : Array(a)) : Maybe(a) { ... }
```

This function is actually type safe, and Mint supports it nicely. But it gets more complex when you want to keep some generics consistency inside a component:

```
component DangerouslyUnsafe {
  property values : Array(a) = []
  property renderValue : Function(a, Html) = (a : a) : Html { Html.empty() }

  fun render : Html {
    <div>
      <{ for (value of values) {
        renderValue(value)
      }}>
    </div>
  }
}

/* later... */

record Deep {
  title : String
}

record Something {
  deep : Deep
}

component Main {
  render : Html {
    <DangerouslyUnsafe
      values=[1, 2, 3]
      renderValue={(value : Something) : Html {
        <div> <{value.deep.title}></div>
      }}
      />
  }
}
```

Will raise an error at runtime: `Uncaught TypeError: Cannot read property 'title' of undefined`.

It could be solve by making components generics:

```

component Safe(a) {
  property values : Array(a) = []
  property renderValue : Function(a, Html) = (a : a) : Html { Html.empty() }

  fun render : Html {
    <div>
      <{ for (value of values) {
        renderValue(value)
      }}>
    </div>
  }
}

/* later... */

record Deep {
  title : String
}

record Something {
  deep : Deep
}

component Main {
  render : Html {
    <Safe
      values=[1, 2, 3]
      renderValue={(value : Something) : Html { /* error at compile time: Number != something */
        <div> <{value.deep.title}></div>
      }}
      />
  }
}
```

It would greatly improve the type safety!

Also, is there any plan to support "type holes/generics" in record and property before v1?

**Comment by gdotdesign** [comment]
This is something I'm thinking about lately as well.

> According to this line: https://github.com/mint-lang/mint/blob/master/src/messages/property_not_hole.cr#L9, "type hole" or "generics" are not yet supported in properties (neither are they supported in records).

This error is not on 0.5 but will be 0.6 so the example with the `DangerouslyUnsafe` component will error out.

> Also, is there any plan to support "type holes/generics" in record and property before v1?

Generics won't be implemented before 1.0 but planned after that in some way.
